tracing/src/main/java/com/palantir/remoting3/tracing/Tracers.java

```diff
@@ -177,19 +171,42 @@ public V call() throws Exception {
      */
     private static class TracingAwareRunnable implements Runnable {
         private final Runnable delegate;
-        private final Trace trace;
+        private DeferredTracer deferredTracer;
 
         TracingAwareRunnable(Runnable delegate) {
             this.delegate = delegate;
-            this.trace = Tracer.copyTrace();
+            this.deferredTracer = new DeferredTracer();
         }
 
         @Override
         public void run() {
+            deferredTracer.withTrace(() -> {
+                delegate.run();
+                return null;
+            });
+        }
+    }
+
+    public interface ThrowingCallable<T, E extends Throwable> {
+        T call() throws E;
+    }
+
+    public static final class DeferredTracer {
```

Comment:
this is great, should likely add a basic javadoc though that shows example usage?